### ISSUE TITLE
Question Data Processing

### ISSUE THREAD

**Issue by mdaehl** [InhwanBae/Crowd-Behavior-Generation#6]
Hello, in your released preprocessed data there are already gt semantic maps available for the different scenes. I cannot find these in the original datasets.
How did you get these? Were they labeled by you afterwards?

**Comment by InhwanBae** [comment]
Hi @mdaehl,

Thanks for your interest in our work! The semantic maps we use are not **ground-truth** maps and therefore are not included in the original datasets.

To generate semantic maps for each dataset scene, we used the pretrained [Grounded SAM 2](https://github.com/IDEA-Research/Grounded-SAM-2) ([grounded_sam2_local_demo.py](https://github.com/IDEA-Research/Grounded-SAM-2/blob/main/grounded_sam2_local_demo.py)). I experimented with several alternatives, including satellite segmentation models, pretrained SegFormer, pretrained segmentation models in the Y-Net trajectory predictor, and our trial models trained from ground-truth segmentation labels. But we found that Grounded SAM 2 produced the most reliable results.

|![Image](https://github.com/user-attachments/assets/ed82cc4a-c3bb-42a7-9196-2c886d63a711)|![Image](https://github.com/user-attachments/assets/87ecb0c8-a356-4a9e-b899-887d3a0e4613)|![Image](https://github.com/user-attachments/assets/d73c86b5-50f7-471c-8a91-3e5858e2d738)|
|:---:|:---:|:---:|
|Y-Net pretrained|Grounded SAM 2|Grounded SAM 2 + Postprocess|

For Grounded SAM 2, we provide text labels, for example,
`text_prompt = "sidewalk. pavement. road. driveway. grass. lawn. terrain. bush. shrub. tree. structure. building. house. wall. car. vehicle. background."` We tested various text prompts and selected the most effective one for each dataset. Afterward, we removed dynamic objects like pedestrians, bicycles and vehicles, and then computed a weighted sum of all masks across 7 classes. Lastly, we assigned each pixel to the class with the maximum response to get the final semantic segmentation map.

If you have any further questions, feel free to reach out!

**Comment by mdaehl** [comment]
Thank you so much for the quick and detailed answer @InhwanBae ! In your paper and also in the README you present a cool example of the evaluation on New York City. Could you please explain how you were able to create these? It would be interesting to know how you both created the input for the model (BEV image) and how you created the respective 3D model for visualization.

**Comment by InhwanBae** [comment]
To apply our CrowdES framework to additional scenes such as New York City, you can use the [Custom Input](https://github.com/InhwanBae/Crowd-Behavior-Generation?tab=readme-ov-file#run-crowdes-with-custom-input) option. Since the README provides only limited details, I explain the process here more thoroughly.



### 1. Crowd Behavior Generation

To test CrowdES on scenes beyond the provided evaluation datasets, you need to add custom data under the `datasets/Synthetic/` directory. Specifically, populate the following subfolders:

* `image_terrain/`: contains the BEV background image. In my experiments, I used satellite or map images from Google Maps.
* `homography/`: contains a text file specifying the scale, i.e., how many meters correspond to one pixel in the image.
* `gt/`: contains maps of walkable regions, which Google Maps provides as layers. These maps include appearance density and population density maps.

The folder structure should look like this:
```
datasets/Synthetic
├─ image_terrain
│  └─ manhattan_bg.png
├─ homography
│  └─ manhattan_scale.txt
└─ gt
   ├─ manhattan_appearance_density.png
   └─ manhattan_population_density.png
```

> [!NOTE]
> For convenience, I share the [**real-world datasets.zip**](https://github.com/InhwanBae/Crowd-Behavior-Generation/releases/tag/v1.0-dataset) file that I used in writing the paper. Since the task involves large-scale outdoor environments, I used the SDD pretrained weights in the paper.



### 2. 3D Visualization

Setting up 3D visualization for custom real-world scenes is somewhat tricky and may take about a day initially (after that, visualizing 3D maps becomes surprisingly easier).

> [!NOTE]
> Due to potential license issues with UE4, I cannot publicly upload the files. If you encounter difficulties following this process, feel free to reach out to me via the personal email address listed in the paper, and I will share the full UnrealEngine4 project files.

#### 2.1. Extracting Maps from Google Earth

**Option 1: Using Microsoft Flight Simulator SDK (new)**
First, download the [Google Earth Decoder](https://drive.google.com/file/d/18zdIjLbRgM5Ce1PtFPKYn-bCfOZcpPAO/view) and [patch files](https://flightsim.to/file/39900/gogole-earth-decoder-update) for MSFS. Next, select the area to capture the 3D maps for download. Then, import the captured `.gltf` data into [Blender](https://www.blender.org/download/), select all meshes, and click `File → External Data → Find Missing Files` to add textures. Finally, export the processed 3D map as an FBX file from Blender. You should embed all textures into a single FBX file by changing the Path Mode to Copy and clicking the Embed Textures icon. A useful tutorial can be found here: [(YouTube) Import Google map 3d models into your 3d software](https://www.youtube.com/watch?v=eug3NGpmv-0)

**Option 2: Using RenderDoc**
First, use [RenderDoc](https://renderdoc.org/) to capture 3D maps from Google Earth. Then, import the captured `.rdc` data into [Blender 4.1](https://download.blender.org/release/Blender4.1/) using [MapsModelsImporter](https://github.com/eliemichel/MapsModelsImporter). Finally, export the processed 3D map as an FBX file from Blender. You should embed all textures into a single FBX file by changing the Path Mode to Copy and clicking the Embed Textures icon. A useful tutorial can be found here: [(YouTube) Google Maps to Blender](https://www.youtube.com/watch?v=hrDW0Gh0tx4).

**Tip**: It would be better to merge all fragmented meshes by pressing `Ctrl + J` in Blender to make it easier to apply textures in Unreal Engine.

#### 2.2. Building UnrealEngine-CARLA

Next, launch the CARLA map editor. Follow the [instructions](https://github.com/carla-simulator/carla/tree/0.9.15.2?tab=readme-ov-file#documentation) to compile UnrealEngine-CARLA. (I used UE4 version, but UE5 will also fine) After compilation, you will be able to access the Unreal Engine map development interface, as shown below:
```
git clone -b 0.9.15.2 https://github.com/carla-simulator/carla.git carla_0.9.15.2
git clone --depth 1 -b carla https://github.com/CarlaUnreal/UnrealEngine.git carla_ue4
```

<img width="3839" height="2075" alt="Image" src="https://github.com/user-attachments/assets/96c0a339-6df5-4256-ba24-5f93b30ccfc7" />

#### 2.3. Loading the 3D Map and Launching Visualization

Start by duplicating an existing Town (e.g., Town 1). Create a plane high enough so that it is not interfered with by buildings, set `Collide = True`, and apply the BEV image as a texture. Adjust the scale so that the map dimensions match real-world measurements. Then, import the FBX model exported from Blender, and align its size and position with the plane. (Optional) Set the FBX model to `Collide = False` and make the plane transparent for better visualization. Make sure to record the **origin x, y coordinates** of the plane.

Finally, for visualizing crowd trajectories in 3D, use the scripts provided in the [3D Visualization Toolkit](https://github.com/InhwanBae/Crowd-Behavior-Generation/tree/main/3D%20Visualization%20Toolkit#-usage). Update the parameters `origin`(top-left meter coordinates in Unreal Engine) and `multi`(pixel-to-meter ratio) in the script according to your new scene. See the relevant code snippet here:

https://github.com/InhwanBae/Crowd-Behavior-Generation/blob/299ea7bba21a2d8ad73d8ea1d5a4a31b058a0482/3D%20Visualization%20Toolkit/run_trajsim.py#L637-L657
